tests/sentry/ratelimits/test_sliding_windows.py

```diff
@@ -0,0 +1,103 @@
+import pytest
+
+from sentry.ratelimits.sliding_windows import (
+    GrantedQuota,
+    Quota,
+    RedisSlidingWindowRateLimiter,
+    RequestedQuota,
+)
+
+
+@pytest.fixture
+def limiter():
+    return RedisSlidingWindowRateLimiter()
+
+
+TIMESTAMP_OFFSET = 100
+
+
+def test_empty_quota(limiter):
+    resp = limiter.check_and_use_quotas(
+        [
+            RequestedQuota(
+                prefix="foo",
+                requested=1,
+                quotas=[
+                    Quota(
+                        window_seconds=10,
+                        granularity_seconds=1,
+                        limit=0,
+                    )
+                ],
+            )
+        ]
+    )
+    assert resp == [GrantedQuota(prefix="foo", granted=0)]
+
+
+def test_basic(limiter):
+    quotas = [
+        Quota(
+            window_seconds=10,
+            granularity_seconds=1,
+            limit=10,
+        )
+    ]
+
+    for timestamp in range(10):
+        resp = limiter.check_and_use_quotas(
```

Comment:
food for thought:

I know that `check_and_use_quotas` just calls `check_within_quotas` and then `use_quotas` but it doesn't seem like it's going to be used this way which makes me wonder if the tests need to change or if the interface needs to be different

e.g.

```python3
with RedisSlidingWindowRateLimiter.check_and_use_quotas(....): # check_within_quotas
    # do stuff
# use quotas at exit
```

I won't be using this library so I can't say for sure but just something to think about

Reply:
`check_and_use_quotas` is only going to be used in tests right now but I put it on the baseclass anyway just because I _could_ imagine it _could_ be useful in other scenarios